first pass at TableAnnotate IR

catoverdrive opened the pull request:
@cseed @danking I tried to do annotate but I think it's pretty messy right now.

cseed commented on src/main/scala/is/hail/expr/Parser.scala:
```diff
@@ -147,6 +147,39 @@ object Parser extends JavaTokenParsers {
+    val asts = new Array[AST](parsed.size)
+
+    var i = 0
+    parsed.foreach { case (name, ast, _) =>
```
You need to handle splat here.  See the discussion on splat here for example: https://hail.is/docs/stable/hail.VariantDataset.html#hail.VariantDataset.export_variants.

tpoterba replied:
can't splat go away in hail2?

cseed commented on src/main/scala/is/hail/expr/Relational.scala:
```diff
@@ -625,4 +628,75 @@ case class TableFilter(child: TableIR, pred: IR) extends TableIR {
   }
+}
+
+case class TableAnnotate(child: TableIR, paths: IndexedSeq[List[String]], preds: IndexedSeq[IR]) extends TableIR {
```
This begs the question what "level" the TableIR nodes should operate at:

1. should they operate at the user level as you've done, that is, the IR nodes roughly correspond to what the user writes for an expression,

2. at the opposite extreme, they could be expressions that consume and produce a singel value (the row), in which case the logic to build the nested stack of `InsertFields` would be outside the IR node and this IR node would essentially correspond to "map",

3. the intermediate solution that treats a table essentially as a struct, and TableAnnotation would essentially correspond to InsertFields, and instead of `paths: ISeq[List[String]]` you'd have `fields: ISeq[String]`.

I'm inclined against (1) since the user interface should be designed to maximize conciseness and expressiveness for the user while the IR should be designed to maximize the ability to be analyzed and optimized.  (3) has a nice symmetry between the struct operators in the IR and the table operators in the TableIR, but I think (2) will ultimately be simpler because we won't have to duplicate logic between the TableIR and IR.

The other nice (?)  thing about (2) is that it forces us to handle the analysis and optimization of structs at every level because the structs at the top level and deeply nested structs are expressed in the same way.  With (3), there's a temptation to give special treatment to the toplevel fields.

What do you think?

cseed commented on src/main/scala/is/hail/expr/Relational.scala:
```diff
@@ -625,4 +628,75 @@ case class TableFilter(child: TableIR, pred: IR) extends TableIR {
+  def execute(hc: HailContext): TableValue = {
+    val tv = child.execute(hc)
+    Infer(newIR)
+    val fb = FunctionBuilder.functionBuilder[Region, Long, Boolean, Long, Boolean, Long]
```
You should use the higher level `Compile` interface here (or modify it so it will work for you.)

catoverdrive commented:
@cseed I moved unwrapping of the nested insert before the TableAnnotate node, as we talked about. The code is pretty messy and I'm still tidying it up, but it's now there if you wanted to take a look.

re: splats---I talked to Tim about this earlier and he didn't think that annotate needed to support them, but I added a test case anyway and it seems to work fine.

danking commented:
@catoverdrive needs a rebase
